draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1149,47 +1159,49 @@ Procedure:
 Case 2: q = 5 (mod 8)
 
 Constants:
-1. c1 = sqrt(-1) in F, i.e., c1 * c1 = -1 mod q.
+1. c1 = sqrt(-1) in F, i.e., (c1 * c1) = -1 in F
 
 Procedure:
 1. t1 = x^((q + 3) / 8)
-2. e  = t1 * t1 == x
-3. s  = CMOV(t1 * c1, t1, e)
+2.  e = (t1 * t1) == x
+3.  s = CMOV(t1 * c1, t1, e)
 3. return s
 
 ======
 
 Case 3: q = 9 (mod 16)
 
 Constants:
-1. c1 = sqrt(-1) in F, i.e., c1 * c1 = -1 mod q.
-2. c2 = sqrt(sqrt(-1)) in F, i.e., c2 * c2 = c1 mod q.
-3. c3 = sqrt(-sqrt(-1)) in F, i.e., c3 * c3 = -c1 mod q.
+1. c1 = sqrt(-1) in F, i.e., (c1 * c1) == -1 in F
+2. c2 = sqrt(c1) in F, i.e., (c2 * c2) == c1 in F
+3. c3 = sqrt(-c1) in F, i.e., (c3 * c3) == -c1 in F
 
 Procedure:
 1.  t1 = x^((q + 7) / 16)
 2.  t2 = c1 * t1
 3.  t3 = c2 * t1
 4.  t4 = c3 * t1
-5.  e1 = t2 * t2 == x
-6.  e2 = t3 * t3 == x
-7.  t1 = CMOV(t1, t2, e1)  // select t2 if t2 * t2 == x
-8.  t2 = CMOV(t4, t3, e2)  // select t3 if t3 * t3 == x
-9.  e3 = t2 * t2 == x
-10. s  = CMOV(t1, t2, e3)  // select the sqrt from t1 and t2
+5.  e1 = (t2 * t2) == x
+6.  e2 = (t3 * t3) == x
+7.  t1 = CMOV(t1, t2, e1)  // select t2 if (t2 * t2) == x
+8.  t2 = CMOV(t4, t3, e2)  // select t3 if (t3 * t3) == x
+9.  e3 = (t2 * t2) == x
+10.  s = CMOV(t1, t2, e3)  // select the sqrt from t1 and t2
 11. return s
 ~~~
 
--   sgn0(x): This function returns either +1 or -1, indicating the sign of x.
-    This function considers 0 to be positive.
+-   sgn0(x): This function returns either +1 or -1 indicating the "sign" of x,
+    where sgn0(x) == -1 just when x is lexically greater than -1 * x.
+    Thus, this function considers 0 to be positive.
     The following procedure implements sgn0(x) in constant time.
     See {{bg-curves}} for a discussion of representing x as a vector.
 
 ~~~
 sgn0(x)
 
 Parameters:
-  1. F, a finite field of characteristic p and order q = p^m, m >= 1.
+- F, a finite field of characteristic p and order q = p^m, m >= 1.
+
 Input: x, an element of F.
 Output: -1 or 1.
```

Comment:
In principle it could be either, right?

Also, in the definition of abs we say `sgn0(x) * x`, which kind of implies interpreting sgn0 as returning an element of F.

Reply:
It does, however, in the implementation side -1 \in ZZ (or even -1 as a boolean) is different from -1 in F.

Reply:
True! Fixed.